cusub_cortex/state_machine/src/tasks/start_gate_det.py

```diff
@@ -0,0 +1,685 @@
+
+        # Find start_gate pole dobject number and check for errors
+        dobj_nums = self.listener.query_class(self.target_class_id)
+        if len(dobj_nums) > 1: # Check if more than 1 instance of target class
+            self.cuprint(str(len(dobj_nums)) + " " + self.target_class_id + " classes detected!", warn=True)
+            self.cuprint("selecting the second", warn=True)
+        elif not dobj_nums: # Check if target class is not present (shouldn't be possible)
+            self.cuprint("somehow no " + self.target_class_id + " classes found?", warn=True)
+            return "lost_object"
+        # DOBJECT NUMBER
+        self.cuprint(str(dobj_nums), warn=True)
+        dobj_num = dobj_nums[1]
+        self.cuprint("located " + self.target_class_id + "'s dobject num: " + str(dobj_num))
+
+        # TODO start a watch dog timer on detections
+
+        # Enable the PID loops
+        self.yaw_client.enable()
+        self.drive_client.enable()
+        self.depth_client.enable()
+        self.drive_client.set_setpoint(self.mag_target)
+        self.depth_client.set_setpoint(-1.5)
+        self.in_position_state = None
+
+        self.cuprint("servoing")
+        r = rospy.Rate(self.rate)
+        while not rospy.is_shutdown():
+            if self.listener.check_new_dv(dobj_num):
+                [az_1, el_1, mag_1] = self.listener.get_avg_bearing(dobj_nums[0], num_dv=5)
+                [az_2, el_2, mag_2] = self.listener.get_avg_bearing(dobj_num, num_dv=5)
+                az = ((az_1 + az_2)/2)
+                el = ((el_1 + el_2)/2)
+                mag = np.sqrt(((mag_1 + mag_2)/2))
+
+                self.yaw_client.set_setpoint(az, loop=False)
+                if abs(self.mag_target - mag) < 0.5*self.mag_target:
+                    self.cuprint("[MAGNITUDE: " + str(mag) + "] - " + "[AZ: " + str(az) + "] - " +  "[CHECK POSITION: " + str(self.count) + " / " + str(self.count_target) + "]")
+                    self.count += 1
+                    if self.count == (self.count_target / 3):
+                        self.in_position_state = self.drive_client.get_standard_state()
+                        self.drive_client.set_setpoint(self.in_position_state, loop=False)
+                    # Stop Sub
+                    if self.count > self.count_target:
+                        self.cuprint("in position")
+                        self.yaw_client.set_setpoint(0, loop=False)
+                        self.drive_client.increment_setpoint(0.15, loop=False)
+                        rospy.sleep(5)
+                        break
+                elif self.mag_target - mag > 0:
+                    self.cuprint("mag: " + str(mag))
+                    # Move sub forward
+                    if self.count:
+                        self.drive_client.increment_setpoint(0.2, loop=False)
+                    else:
+                        self.drive_client.increment_setpoint(0.6, loop=False)
+                else:
```

Comment:
since the general process of this looks fairly similar between all of the Approach_X_poles' execute method. I bet we could cut down on the code quite a bit by having just one class and instantiating it three times and calling the init method with different parameters. Or take the approach of the droppers where Drop Objective inherits from Approach since their execute() method is fairly similar. Either method might lead to a cleaner and more compact design